app/src/main/cpp/skyline/nce.cpp

```diff
@@ -1,231 +1,319 @@
-            raw += fmt::format("{:08X}", instr);
+    void NCE::ThreadTrace(u16 numHist, ThreadContext *ctx) {
+        std::string raw;
+        std::string trace;
+        std::string regStr;
+        ctx = ctx ? ctx : state.ctx;
+        if (numHist) {
+            std::vector<u32> instrs(numHist);
+            u64 size = sizeof(u32) * numHist;
+            u64 offset = ctx->pc - size + (2 * sizeof(u32));
+            state.process->ReadMemory(instrs.data(), offset, size);
+            for (auto &instr : instrs) {
+                instr = __builtin_bswap32(instr);
+                if (offset == ctx->pc)
+                    trace += fmt::format("\n-> 0x{:X} : 0x{:08X}", offset, instr);
+                else
+                    trace += fmt::format("\n   0x{:X} : 0x{:08X}", offset, instr);
+                raw += fmt::format("{:08X}", instr);
+                offset += sizeof(u32);
+            }
         }
-        state.logger->Debug("Raw Instructions: 0x{}", raw);
-        state.logger->Debug("CPU Context:");
-        state.logger->Debug("SP: 0x{:X}", regs.sp);
-        state.logger->Debug("PSTATE: 0x{:X}", regs.pstate);
-        for (u16 index = 0; index < constant::NumRegs - 2; index++) {
-            state.logger->Debug("X{}: 0x{:X}", index, regs.regs[index]);
+        if (ctx->faultAddress)
+            regStr += fmt::format("\nFault Address: 0x{:X}", ctx->faultAddress);
+        if (ctx->sp)
+            regStr += fmt::format("\nStack Pointer: 0x{:X}", ctx->sp);
+        for (u16 index = 0; index < constant::NumRegs - 1; index += 2) {
+            auto xStr = index < 10 ? " X" : "X";
+            regStr += fmt::format("\n{}{}: 0x{:<16X} {}{}: 0x{:X}", xStr, index, ctx->registers.regs[index], xStr, index + 1, ctx->registers.regs[index + 1]);
         }
+        if (numHist) {
+            state.logger->Debug("Process Trace:{}", trace);
+            state.logger->Debug("Raw Instructions: 0x{}", raw);
+            state.logger->Debug("CPU Context:{}", regStr);
+        } else
+            state.logger->Debug("CPU Context:{}", regStr);
     }
 
-    u64 NCE::GetRegister(Xreg regId, pid_t pid) {
-        return registerMap.at(pid ? pid : currPid).regs[static_cast<uint>(regId)];
-    }
+    const std::array<u32, 16> CntpctEl0 = {
+        0xD10083FF, // SUB SP, SP, #32
+        0xA90107E0, // STP X0, X1, [SP, #16]
+        0xD28F0860, // MOV X0, #30787
+        0xF2AE3680, // MOVK X0, #29108, LSL #16
+        0xD53BE001, // MRS X1, CNTFRQ_EL0
+        0xF2CB5880, // MOVK X0, #23236, LSL #32
+        0xD345FC21, // LSR X1, X1, #5
+        0xF2E14F80, // MOVK X0, #2684, LSL #48
+        0x9BC07C21, // UMULH X1, X1, X0
+        0xD347FC21, // LSR X1, X1, #7
+        0xD53BE040, // MRS X0, CNTVCT_EL0
+        0x9AC10801, // UDIV X1, X0, X1
+        0x8B010421, // ADD X1, X1, X1, LSL #1
+        0xD37AE420, // LSL X0, X1, #6
+        0xF90003E0, // STR X0, [SP, #0]
+        0xA94107E0, // LDP X0, X1, [SP, #16]
+    };
```

Comment:
Any reason for not including this in guest.S?

Reply:
Not really, I suppose it's a good idea to move it in there.